Yoast/wordpress-seo
# DUPP-41 configuration workout routes

## Context


* We want to add a set of routes to allow the configuration workout to set options, and a set of filters to allow the add-ons to hook into for their own workouts.

## Summary


This PR can be summarized in the following changelog entry:

* Introduces a set of routes for the configuration workout and a series of filters for the addons to add their own workouts stuff.

## Relevant technical choices:

*

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Test the new routes for the configuration workout

* install the JSON Basic Authentication plugin
* choose a user with role `admin` or `SEO Manager` (having the `wpseo_manage_options` capability)
* use Insomnia or Postman to send POST request to the endpoints. Using the plugin above allows you to user the `Basic` authentication, so you can provide the username and password of the user you've chosen.

Endpoints list:
* `/wp-json/yoast/v1/workouts/site_representation` with JSON payload having the format:
```
{
    "company_or_person": "company",
    "company_name": "Acme Inc.",
    "company_logo": "https:\/\/acme.com\/someimage.jpg",
    "company_logo_id": 123,
    "person_logo": "https:\/\/acme.com\/someimage.jpg",
    "person_logo_id": 123,
    "company_or_person_user_id": 321,
    "description": "A nice tagline"
}
```

* `/wp-json/yoast/v1/workouts/social_profiles` with JSON payload having the format:
```
{
    "facebook_site": "https:\/\/facebook.com\/somepage",
    "twitter_site": "somenick",
    "instagram_url": "https:\/\/instagram.com\/somepage",
    "linkedin_url": "https:\/\/linked.in\/somepage",
    "myspace_url": "https:\/\/myspace.com\/somepage",
    "pinterest_url": "https:\/\/pinterest.com\/somepage",
    "youtube_url": "https:\/\/youtube.com\/somepage",
    "wikipedia_url": "https:\/\/en.wikipedia.org\/somepage"
}
```


* `/wp-json/yoast/v1/workouts/enable_tracking` with JSON payload having the format:
```
{ "tracking" : true }
```

* Test changing the values, like using non-numeric strings where an integer is expected, or using something different from `company` or `person`, or not submitting a boolean to the traking endpoint, so you see the validation working

* check that the options are changed in the DB either inspecting the `wpseo_titles` and `wpseo_social` options (and `wpseo_options` for tracking), or visiting the UI (e.g. Search appearance)

#### Test the moving of the Premium stuff
* Without Premium active, visit the Workouts page and check that it's working (without errors, even in the console)
* install the branch from https://github.com/Yoast/wordpress-seo-premium/pull/3497 and activate Premium
* check that the existing workouts work as expected.

### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

* NA the full feature will be tested

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Other environments


## Documentation


## Quality assurance


Fixes [DUPP-41]


[DUPP-41]: https://yoast.atlassian.net/browse/DUPP-41?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
